File: packages/python/src/opensms/resources/otp.py
"""The ``otp`` resource."""

from __future__ import annotations

from typing import Optional

from .._serialize import prune
from ..models import OtpSendResult, OtpVerifyResult
from ._base import Resource


class Otp(Resource):
    """One-time passcodes. Accessed as ``client.otp``."""

    def send(
        self,
        *,
        to: str,
        sender_id: Optional[str] = None,
        template: Optional[str] = None,
        length: Optional[int] = None,
        ttl_seconds: Optional[int] = None,
        idempotency_key: Optional[str] = None,
    ) -> OtpSendResult:
        """Generate and send a code. ``template`` must contain ``{{code}}``."""
        body = prune(
            {
                "to": to,
                "sender_id": sender_id,
                "template": template,
                "length": length,
                "ttl_seconds": ttl_seconds,
            }
        )
        return self._t.request(
            "POST", "/v1/otp/send", json_body=body, idempotent=True, idempotency_key=idempotency_key
        )

    def verify(self, *, otp_id: str, code: str) -> OtpVerifyResult:
        """Check a code. A wrong code returns ``valid: False`` and uses up an
        attempt, so this call is never retried."""
        return self._t.request("POST", "/v1/otp/verify", json_body={"otp_id": otp_id, "code": code})

File: packages/ruby/lib/opensms/resources/otp.rb
# frozen_string_literal: true

require_relative "base"

module Opensms
  module Resources
    # The +otp+ resource: send and verify one-time passcodes. Accessed as
    # +client.otp+.
    class Otp < Base
      # Send a code. POST /v1/otp/send -> 201 { otp_id: }.
      #
      # @param params [Hash] :to (required), :sender_id, :template (must contain
      #   "{{code}}"), :length (4..10), :ttl_seconds (30..86400)
      # @return [Hash] { otp_id: }
      def send(params = nil, idempotency_key: nil, **kwargs)
        body = Models.build(merge(params, kwargs), %i[to sender_id template length ttl_seconds], required: %i[to])
        @transport.request(:post, "/v1/otp/send", body: body, idempotency_key: idem(idempotency_key))
      end

      # Check a code. A wrong code returns { valid: false } and burns an
      # attempt, so this call is never auto-retried.
      # POST /v1/otp/verify -> { valid:, attempts_left: }.
      #
      # @param params [Hash] :otp_id, :code
      # @return [Hash]
      def verify(params = nil, **kwargs)
        body = Models.build(merge(params, kwargs), %i[otp_id code], required: %i[otp_id code])
        @transport.request(:post, "/v1/otp/verify", body: body)
      end
    end
  end
end
